File: Kakao/2019인턴/hotel.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

unordered_map<long long,long long> map;
vector<long long> answer;
long long K;


long long getRoom(long long room){

    if(map.find(room) == map.end()){
        
        map[room] = (room+1) % (K+1);
        answer.push_back(room);
        return map[room];
    }
    else{
        return map[room] = getRoom(map[room]);
    }

}


vector<long long> solution(long long k, vector<long long> room_number) {
    K = k;

    for(int i =0;i<room_number.size();i++){
        map[room_number[i]] = getRoom(room_number[i]);
    }


    return answer;
}

File: Kakao/2019인턴/hotel.js
let answer = [];
let map = new Map();

function checkIn(wantRoom){
    let visitedRoom = [];
    while(map.get(wantRoom)){
        visitedRoom.push(wantRoom);
        wantRoom = map.get(wantRoom);
    }
    visitedRoom.push(wantRoom);
    for(let i =0;i<visitedRoom.length;i++){
        map.set(visitedRoom[i], wantRoom+1);
    }
    return wantRoom;

}

function solution(k, room_number) {
    
    

    for(let i =0;i<room_number.length;i++){
        let wantRoom = room_number[i];
        answer[i] = checkIn(wantRoom);
    }


    return answer;
}


(function solve(){
    let k = 10;
    let rooms = [1,3,4,1,3,1];
    
    console.log(solution(k,rooms));

})();